core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala

```diff
@@ -201,12 +203,20 @@ private[spark] class HeartbeatReceiver(sc: SparkContext, clock: Clock)
           s"${now - lastSeenMs} ms exceeds timeout $executorTimeoutMs ms")
         scheduler.executorLost(executorId, SlaveLost("Executor heartbeat " +
           s"timed out after ${now - lastSeenMs} ms"))
-          // Asynchronously kill the executor to avoid blocking the current thread
+        // Asynchronously kill the executor to avoid blocking the current thread
         killExecutorThread.submit(new Runnable {
           override def run(): Unit = Utils.tryLogNonFatalError {
             // Note: we want to get an executor back after expiring this one,
             // so do not simply call `sc.killExecutor` here (SPARK-8119)
             sc.killAndReplaceExecutor(executorId)
+            // In case of the executors which are not gracefully shut down, we should remove
+            // lost executors from CoarseGrainedSchedulerBackend manually here (SPARK-27348)
+            sc.schedulerBackend match {
```

Comment:
Why not put this into TaskSchedulerImpl.executorLost() ?

Reply:
I think that causes a loop / cycle because `CoarseGrainedSchedulerBackend.removeExecutor` directly calls `TaskSchedulerImpl.executorLost`.

Reply:
ah, i see